py/server/deephaven/table.py

```diff
@@ -43,6 +45,43 @@
 _JReverseAsOfJoinRule = jpy.get_type("io.deephaven.api.ReverseAsOfJoinRule")
 _JTableOperations = jpy.get_type("io.deephaven.api.TableOperations")
 
+# Dynamic Query Scope
+_JQueryScope = jpy.get_type("io.deephaven.engine.table.lang.QueryScope")
+_JUnsynchronizedScriptSessionQueryScope = jpy.get_type(
+    "io.deephaven.engine.util.AbstractScriptSession$UnsynchronizedScriptSessionQueryScope")
+_JPythonScriptSession = jpy.get_type("io.deephaven.integrations.python.PythonDeephavenSession")
+_j_script_session = jpy.cast(_JQueryScope.getScope(), _JUnsynchronizedScriptSessionQueryScope).scriptSession()
+_j_py_script_session = jpy.cast(_j_script_session, _JPythonScriptSession)
+_deephaven_pkg_prefix = __file__
+
+
+@contextlib.contextmanager
+def _query_scope_ctx():
+    """A context manager to set/unset query scope based on the scope of the most immediate application code that invokes
+    Table operations."""
+
+    # locate the innermost Deephaven frame (i.e. any of the table operation methods that use this context manager)
```

Comment:
Why do we want the innermost Deephaven frame, rather than the outermost Deephaven-only frame? I'm not sure I understand.

Reply:
The next comment in the code tries to explain it. 
```
   # we deem any module/function/method which calls a wrapped formula/filter-evaluating Table op as an application
   # frame. This holds true and makes sense even when such module/function/method is part of the Deephaven package, and
   # should not be a hurdle for developing such code as function/method parameters are now part of the query scope
```
Basically we wan to find the wrapper function that directly uses this context manager and uses its caller's locals and the module globals as the query scope. I have changed the names from application/app to caller in the code to make it clear. This is less ambiguous than the notion of 'application scope', which in most cases is the same as the immediate caller's.